server/util/log/gcp/gcp.go

```diff
@@ -0,0 +1,122 @@
+package gcp
+
+import (
+	"context"
+	"encoding/json"
+	"flag"
+	"strconv"
+	"time"
+
+	"cloud.google.com/go/logging"
+	"github.com/rs/zerolog"
+
+	logpb "google.golang.org/genproto/googleapis/logging/v2"
+	structpb "google.golang.org/protobuf/types/known/structpb"
+)
+
+var (
+	ProjectID = flag.String("app.log_gcp_project_id", "", "The project ID to log to in GCP (if any).")
+	LogID     = flag.String("app.log_gcp_log_id", "", "The log ID to log to in GCP (if any).")
+)
+
+func NewLogWriter() (zerolog.LevelWriter, error) {
+	if *ProjectID == "" || *LogID == "" {
+		return nil, nil
+	}
+	client, err := logging.NewClient(context.Background(), *ProjectID)
+	if err != nil {
+		return nil, err
+	}
+	return &logWriter{ctx: context.Background(), logger: client.Logger(*LogID)}, nil
+}
+
+type logWriter struct {
+	ctx    context.Context
+	logger *logging.Logger
+}
+
+func (l *logWriter) Write(p []byte) (int, error) {
+	l.logger.StandardLogger(logging.Default).Print(string(p))
+	return len(p), nil
+}
+
+func (l *logWriter) WriteLevel(level zerolog.Level, p []byte) (int, error) {
+	entry := logging.Entry{}
+	m := map[string]any{}
+	if err := json.Unmarshal(p, &m); err == nil {
+		jsonPayload := &structpb.Struct{Fields: map[string]*structpb.Value{}}
+		entry.Payload = jsonPayload
+		entry.Operation = &logpb.LogEntryOperation{Producer: "github.com/buildbuddy-io/buildbuddy"}
+		if v, ok := m[zerolog.TimestampFieldName]; ok {
+			if t, ok := v.(string); ok {
+				if entry.Timestamp, err = time.Parse(zerolog.TimeFieldFormat, t); err != nil {
+					entry.Timestamp = time.Time{}
+				}
+				jsonPayload.Fields["timestamp"] = structpb.NewStringValue(t)
+			}
+		}
+		if v, ok := m[zerolog.MessageFieldName]; ok {
+			if p, ok := v.(string); ok {
+				jsonPayload.Fields["message"] = structpb.NewStringValue(p)
+			}
+		}
+		if v, ok := m["executor_id"]; ok {
```

Comment:
Let's avoid hardcoding these fields. Our logging code makes it easy to attach new fields.

You can look at the zerolog builtin logger for a reference on how they handle log fields: https://github.com/rs/zerolog/blob/3543e9d94bc5ed088dd2d9ad1d19c7ccd0fa65f5/console.go#L102